File: src/features/accounts/types/account-form.ts
import type { Translate } from "../../../i18n/context"
import type { AccountSummary } from "../../../lib/supabase/types"
import { multiplyDecimals } from "../../../lib/financial-calculations/decimal"

export const accountTypeOptions = [
  { value: "cash", labelKey: "accountType.cash", creationLabelKey: "accounts.type.cash" },
  { value: "bank", labelKey: "accountType.bank", creationLabelKey: "accounts.type.bank" },
  { value: "brokerage", labelKey: "accountType.brokerage", creationLabelKey: "accounts.type.brokerage" },
  { value: "gold", labelKey: "accountType.gold", creationLabelKey: "accounts.type.gold" },
  { value: "real_estate", labelKey: "accountType.real_estate", creationLabelKey: "accounts.type.realEstate" },
  { value: "business", labelKey: "accountType.business", creationLabelKey: "accounts.type.business" },
  { value: "other", labelKey: "accountType.other", creationLabelKey: "accounts.type.other" },
] as const

export const currencyOptions = [
  { value: "USD", labelKey: "currency.USD" },
  { value: "SAR", labelKey: "currency.SAR" },
  { value: "EGP", labelKey: "currency.EGP" },
  { value: "EUR", labelKey: "currency.EUR" },
  { value: "GBP", labelKey: "currency.GBP" },
] as const

export const bankSubtypeOptions = [
  { value: "debit", labelKey: "accounts.form.bankSubtype.debit" },
  { value: "credit", labelKey: "accounts.form.bankSubtype.credit" },
] as const

export const investmentTypeOptions = [
  { value: "stock_etf", labelKey: "accounts.form.investmentType.stockEtf" },
  { value: "crypto", labelKey: "accounts.form.investmentType.crypto" },
  { value: "other", labelKey: "accounts.form.investmentType.other" },
] as const

export const propertyTypeOptions = [
  { value: "apartment", labelKey: "accounts.form.propertyType.apartment" },
  { value: "villa", labelKey: "accounts.form.propertyType.villa" },
  { value: "land", labelKey: "accounts.form.propertyType.land" },
  { value: "office", labelKey: "accounts.form.propertyType.office" },
  { value: "other", labelKey: "accounts.form.propertyType.other" },
] as const

export const metalTypeOptions = [
  { value: "gold", labelKey: "accounts.form.metalType.gold" },
  { value: "silver", labelKey: "accounts.form.metalType.silver" },
] as const

export const goldPurityOptions = [
  { value: "24k", label: "24k" },
  { value: "22k", label: "22k" },
  { value: "21k", label: "21k" },
  { value: "18k", label: "18k" },
  { value: "14k", label: "14k" },
  { value: "10k", label: "10k" },
  { value: "9k", label: "9k" },
] as const

export const silverPurityOptions = [
  { value: "999", label: "999" },
  { value: "958", label: "958" },
  { value: "950", label: "950" },
  { value: "925", label: "925" },
  { value: "900", label: "900" },
  { value: "835", label: "835" },
  { value: "800", label: "800" },
] as const

export const purityOtherOption = { value: "other", labelKey: "accounts.form.purity.other" } as const

export const accountTypeCodes = accountTypeOptions.map(
  (option) => option.value,
) as [
  (typeof accountTypeOptions)[number]["value"],
  ...(typeof accountTypeOptions)[number]["value"][],
]

export const currencyCodes = currencyOptions.map(
  (option) => option.value,
) as [
  (typeof currencyOptions)[number]["value"],
  ...(typeof currencyOptions)[number]["value"][],
]

export const bankSubtypeCodes = bankSubtypeOptions.map(
  (option) => option.value,
) as [
  (typeof bankSubtypeOptions)[number]["value"],
  ...(typeof bankSubtypeOptions)[number]["value"][],
]

export const investmentTypeCodes = investmentTypeOptions.map(
  (option) => option.value,
) as [
  (typeof investmentTypeOptions)[number]["value"],
  ...(typeof investmentTypeOptions)[number]["value"][],
]

export const propertyTypeCodes = propertyTypeOptions.map(
  (option) => option.value,
) as [
  (typeof propertyTypeOptions)[number]["value"],
  ...(typeof propertyTypeOptions)[number]["value"][],
]

export const metalTypeCodes = metalTypeOptions.map(
  (option) => option.value,
) as [
  (typeof metalTypeOptions)[number]["value"],
  ...(typeof metalTypeOptions)[number]["value"][],
]

export const goldPurityCodes = goldPurityOptions.map((option) => option.value)
export const silverPurityCodes = silverPurityOptions.map((option) => option.value)
export const purityCodes = [
  ...goldPurityCodes,
  ...silverPurityCodes,
  purityOtherOption.value,
] as unknown as [string, ...string[]]

export function getPurityOptions(
  metalType: (typeof metalTypeCodes)[number] | "",
): ReadonlyArray<{ value: string; label?: string; labelKey?: string }> {
  if (metalType === "gold") return [...goldPurityOptions, purityOtherOption]
  if (metalType === "silver") return [...silverPurityOptions, purityOtherOption]
  return []
}

export type AccountTypeCode = (typeof accountTypeCodes)[number]

export type AccountFormValues = {
  name: string
  accountTypeCode: AccountTypeCode
  currencyCode: (typeof currencyCodes)[number]
  openingBalance: string
  bankSubtype: (typeof bankSubtypeCodes)[number] | ""
  investmentType: (typeof investmentTypeCodes)[number] | ""
  balanceGrams: string
  propertyType: (typeof propertyTypeCodes)[number] | ""
  ownershipPercentage: string
  businessType: string
  industry: string
  metalType: (typeof metalTypeCodes)[number] | ""
  purity: string
  purchaseDate: string
  costPerUnit: string
  notes: string
  isActive: boolean
}

export const emptyAccountFormValues: AccountFormValues = {
  name: "",
  accountTypeCode: "cash",
  currencyCode: "USD",
  openingBalance: "0",
  bankSubtype: "",
  investmentType: "",
  balanceGrams: "0",
  propertyType: "",
  ownershipPercentage: "100",
  businessType: "",
  industry: "",
  metalType: "",
  purity: "",
  purchaseDate: "",
  costPerUnit: "0",
  notes: "",
  isActive: true,
}

export function accountToFormValues(
  account: AccountSummary,
): AccountFormValues {
  return {
    name: account.name,
    accountTypeCode: account.account_type_code as AccountFormValues["accountTypeCode"],
    currencyCode: account.currency_code as AccountFormValues["currencyCode"],
    openingBalance: account.opening_balance,
    bankSubtype: (account.bank_subtype ?? "") as AccountFormValues["bankSubtype"],
    investmentType: (account.investment_type ?? "") as AccountFormValues["investmentType"],
    balanceGrams: account.balance_grams ?? "0",
    propertyType: (account.property_type ?? "") as AccountFormValues["propertyType"],
    ownershipPercentage: account.ownership_percentage ?? "100",
    businessType: account.business_type ?? "",
    industry: account.industry ?? "",
    metalType: (account.metal_type ?? "") as AccountFormValues["metalType"],
    purity: account.purity ?? "",
    purchaseDate: account.purchase_date ?? "",
    costPerUnit: account.cost_per_unit ?? "0",
    notes: account.notes ?? "",
    isActive: account.is_active,
  }
}

export function getTotalMetalAmount(values: AccountFormValues): string | null {
  return multiplyDecimals(values.costPerUnit.trim() || "0", values.balanceGrams.trim() || "0")
}

export function getBalanceLabelKey(
  code: AccountFormValues["accountTypeCode"],
): "accounts.form.startingBalance" | "accounts.form.startingCashBalance" | "accounts.form.currentValue" {
  if (code === "brokerage") return "accounts.form.startingCashBalance"
  if (code === "real_estate" || code === "business") return "accounts.form.currentValue"
  return "accounts.form.startingBalance"
}

export function getAccountTypeLabel(code: string, t: Translate): string {
  const option = accountTypeOptions.find((item) => item.value === code)
  return option ? t(option.labelKey) : code.replaceAll("_", " ")
}

export type AccountTypeSpecificFields = {
  openingBalance: string | undefined
  bankSubtype: "debit" | "credit" | null
  investmentType: "stock_etf" | "crypto" | "other" | null
  balanceGrams: string | null
  propertyType: "apartment" | "villa" | "land" | "office" | "other" | null
  ownershipPercentage: string | null
  businessType: string | null
  industry: string | null
  metalType: "gold" | "silver" | null
  purity: string | null
  purchaseDate: string | null
  costPerUnit: string | null
}

export function toAccountTypeSpecificFields(
  values: AccountFormValues,
): AccountTypeSpecificFields {
  const fields: AccountTypeSpecificFields = {
    openingBalance: undefined,
    bankSubtype: null,
    investmentType: null,
    balanceGrams: null,
    propertyType: null,
    ownershipPercentage: null,
    businessType: null,
    industry: null,
    metalType: null,
    purity: null,
    purchaseDate: null,
    costPerUnit: null,
  }

  switch (values.accountTypeCode) {
    case "cash":
    case "other":
      fields.openingBalance = values.openingBalance.trim()
      break
    case "bank":
      fields.openingBalance = values.openingBalance.trim()
      fields.bankSubtype = values.bankSubtype || null
      break
    case "brokerage":
      fields.openingBalance = values.openingBalance.trim()
      fields.investmentType = values.investmentType || null
      break
    case "gold":
      fields.openingBalance = getTotalMetalAmount(values) ?? "0"
      fields.balanceGrams = values.balanceGrams.trim()
      fields.metalType = values.metalType || null
      fields.purity = values.purity.trim() || null
      fields.purchaseDate = values.purchaseDate.trim() || null
      fields.costPerUnit = values.costPerUnit.trim()
      break
    case "real_estate":
      fields.openingBalance = values.openingBalance.trim()
      fields.propertyType = values.propertyType || null
      fields.ownershipPercentage = values.ownershipPercentage.trim()
      break
    case "business":
      fields.openingBalance = values.openingBalance.trim()
      fields.businessType = values.businessType.trim() || null
      fields.industry = values.industry.trim() || null
      fields.ownershipPercentage = values.ownershipPercentage.trim()
      break
  }

  return fields
}

File: apps/mobile/lib/accounts/account_form.dart
// Account form value shape + type-specific field projection — port of the web
// `AccountFormValues` / `toAccountTypeSpecificFields` / `accountToFormValues`
// (src/features/accounts/types/account-form.ts, docs/accounts.md §2.5).

import '../core/decimals.dart';
import 'account_models.dart';

const bankSubtypeCodes = ['debit', 'credit'];
const metalTypeCodes = ['gold', 'silver'];
const investmentTypeCodes = ['stock_etf', 'crypto', 'other'];
const propertyTypeCodes = ['apartment', 'villa', 'land', 'office', 'other'];
const businessTypeCodes = [
  'sole_proprietorship',
  'partnership',
  'llc',
  'private_company',
  'family_owned',
  'other',
];
const industryCodes = [
  'retail',
  'ecommerce',
  'food_beverage',
  'cosmetics_beauty',
  'healthcare',
  'technology',
  'real_estate',
  'construction',
  'manufacturing',
  'education',
  'professional_services',
  'financial_services',
  'logistics_transportation',
  'hospitality_tourism',
  'agriculture',
  'other',
];
const goldPurityCodes = ['24k', '22k', '21k', '18k', '14k', '10k', '9k'];
const silverPurityCodes = ['999', '958', '950', '925', '900', '835', '800'];

List<String> purityOptionsFor(String metalType) {
  if (metalType == 'gold') return [...goldPurityCodes, 'other'];
  if (metalType == 'silver') return [...silverPurityCodes, 'other'];
  return const [];
}

String humanLabel(String code) => code
    .split(RegExp(r'[_\s]+'))
    .where((w) => w.isNotEmpty)
    .map((w) => '${w[0].toUpperCase()}${w.substring(1)}')
    .join(' ');

// Display labels — 1:1 with the web i18n `accounts.form.*` strings so the mobile
// pickers read identically to the web `<select>` options.

const bankSubtypeLabels = {'debit': 'Debit', 'credit': 'Credit'};

const investmentTypeLabels = {
  'stock_etf': 'Stock & ETF',
  'crypto': 'Crypto',
  'other': 'Other',
};

const propertyTypeLabels = {
  'apartment': 'Apartment',
  'villa': 'Villa',
  'land': 'Land',
  'office': 'Office',
  'other': 'Other',
};

const businessTypeLabels = {
  'sole_proprietorship': 'Sole Proprietorship',
  'partnership': 'Partnership',
  'llc': 'Limited Liability Company (LLC)',
  'private_company': 'Private Company',
  'family_owned': 'Family-Owned Business',
  'other': 'Other',
};

const industryLabels = {
  'retail': 'Retail',
  'ecommerce': 'E-commerce',
  'food_beverage': 'Food & Beverage',
  'cosmetics_beauty': 'Cosmetics & Beauty',
  'healthcare': 'Healthcare',
  'technology': 'Technology',
  'real_estate': 'Real Estate',
  'construction': 'Construction',
  'manufacturing': 'Manufacturing',
  'education': 'Education',
  'professional_services': 'Professional Services',
  'financial_services': 'Financial Services',
  'logistics_transportation': 'Logistics & Transportation',
  'hospitality_tourism': 'Hospitality & Tourism',
  'agriculture': 'Agriculture',
  'other': 'Other',
};

const metalTypeLabels = {'gold': 'Gold', 'silver': 'Silver'};

const currencyLabels = {
  'USD': 'USD — US Dollar',
  'SAR': 'SAR — Saudi Riyal',
  'EGP': 'EGP — Egyptian Pound',
  'EUR': 'EUR — Euro',
  'GBP': 'GBP — British Pound',
};

/// Balance-field label by type (web `getBalanceLabelKey`).
String balanceLabelFor(AccountType type) => switch (type) {
  AccountType.brokerage => 'Starting cash balance',
  AccountType.realEstate || AccountType.business => 'Current value',
  _ => 'Current balance',
};

// ---- classification "other:<text>" encoding (businessType / industry) --------

const _otherPrefix = 'other:';

({String value, String custom}) parseClassification(
  String? stored,
  List<String> options,
) {
  if (stored == null || stored.isEmpty) return (value: '', custom: '');
  if (stored.startsWith(_otherPrefix)) {
    return (value: 'other', custom: stored.substring(_otherPrefix.length));
  }
  if (options.contains(stored)) return (value: stored, custom: '');
  return (value: 'other', custom: stored);
}

String? storeClassification(String value, String custom) {
  final v = value.trim();
  if (v.isEmpty) return null;
  if (v != 'other') return v;
  final c = custom.trim();
  return c.isEmpty ? null : '$_otherPrefix$c';
}

String classificationLabel(String? stored, List<String> options) {
  final parsed = parseClassification(stored, options);
  if (parsed.value == 'other') {
    return parsed.custom.isEmpty ? 'Other' : parsed.custom;
  }
  return parsed.value.isEmpty ? '—' : humanLabel(parsed.value);
}

// ---- form values -----------------------------------------------------------

class AccountFormValues {
  AccountFormValues({
    this.name = '',
    this.accountTypeCode = 'cash',
    this.currencyCode = 'USD',
    this.openingBalance = '0',
    this.bankSubtype = '',
    this.creditCardLimit = '',
    this.dueDayOfMonth = '',
    this.investmentType = '',
    this.balanceGrams = '0',
    this.propertyType = '',
    this.ownershipPercentage = '100',
    this.businessType = '',
    this.businessTypeOther = '',
    this.industry = '',
    this.industryOther = '',
    this.location = '',
    String? valuationDate,
    this.valuationMethod = '',
    this.valuationNotes = '',
    this.metalType = '',
    this.purity = '',
    this.purchaseDate = '',
    this.costPerUnit = '0',
    this.notes = '',
    this.isActive = true,
  }) : valuationDate = valuationDate ?? _today();

  String name;
  String accountTypeCode;
  String currencyCode;
  String openingBalance;
  String bankSubtype;
  String creditCardLimit;
  String dueDayOfMonth;
  String investmentType;
  String balanceGrams;
  String propertyType;
  String ownershipPercentage;
  String businessType;
  String businessTypeOther;
  String industry;
  String industryOther;
  String location;
  String valuationDate;
  String valuationMethod;
  String valuationNotes;
  String metalType;
  String purity;
  String purchaseDate;
  String costPerUnit;
  String notes;
  bool isActive;

  AccountType get type => AccountType.fromCode(accountTypeCode);

  static String _today() => DateTime.now().toIso8601String().substring(0, 10);

  static AccountFormValues fromAccount(Account a) {
    final bt = parseClassification(a.businessType, businessTypeCodes);
    final ind = parseClassification(a.industry, industryCodes);
    return AccountFormValues(
      name: a.name,
      accountTypeCode: a.type.code,
      currencyCode: a.currencyCode,
      openingBalance: a.openingBalance,
      bankSubtype: a.bankSubtype ?? '',
      creditCardLimit: a.creditCardLimit ?? '',
      dueDayOfMonth: a.dueDayOfMonth?.toString() ?? '',
      investmentType: a.investmentType ?? '',
      balanceGrams: a.balanceGrams ?? '0',
      propertyType: a.propertyType ?? '',
      ownershipPercentage: a.ownershipPercentage ?? '100',
      businessType: bt.value,
      businessTypeOther: bt.custom,
      industry: ind.value,
      industryOther: ind.custom,
      location: a.location ?? '',
      valuationDate: '',
      metalType: a.metalType ?? '',
      purity: a.purity ?? '',
      purchaseDate: a.purchaseDate ?? '',
      costPerUnit: a.costPerUnit ?? '0',
      notes: a.notes ?? '',
      isActive: a.isActive,
    );
  }
}

/// Available credit for a Bank Credit account is `limit - currentBalance`
/// (web `getCreditCardAmountDue`).
String? creditCardAmountDue(String limit, String currentBalance) =>
    D.subtract(limit.trim(), currentBalance.trim());

/// The camelCase payload columns for `createAccount` / `updateAccount`. Mirrors
/// the web `toAccountTypeSpecificFields`: irrelevant fields are nulled per type,
/// gold's `openingBalance` is forced to "0".
class AccountTypeSpecificFields {
  String? openingBalance;
  String? bankSubtype;
  String? creditCardLimit;
  int? dueDayOfMonth;
  String? investmentType;
  String? balanceGrams;
  String? propertyType;
  String? ownershipPercentage;
  String? businessType;
  String? industry;
  String? location;
  String? metalType;
  String? purity;
  String? purchaseDate;
  String? costPerUnit;

  Map<String, dynamic> toColumns({required String name, String? currencyCode}) {
    final map = <String, dynamic>{
      'name': name,
      'bank_subtype': bankSubtype,
      'credit_card_limit': creditCardLimit,
      'due_day_of_month': dueDayOfMonth,
      'investment_type': investmentType,
      'balance_grams': balanceGrams,
      'property_type': propertyType,
      'ownership_percentage': ownershipPercentage,
      'business_type': businessType,
      'industry': industry,
      'location': location,
      'metal_type': metalType,
      'purity': purity,
      'purchase_date': purchaseDate,
      'cost_per_unit': costPerUnit,
    };
    if (currencyCode != null) map['currency_code'] = currencyCode;
    if (openingBalance != null) map['opening_balance'] = openingBalance;
    return map;
  }
}

AccountTypeSpecificFields toAccountTypeSpecificFields(AccountFormValues v) {
  final f = AccountTypeSpecificFields();
  switch (v.type) {
    case AccountType.cash:
    case AccountType.other:
      f.openingBalance = v.openingBalance.trim();
    case AccountType.bank:
      f.openingBalance = v.openingBalance.trim();
      f.bankSubtype = v.bankSubtype.isEmpty ? null : v.bankSubtype;
      if (v.bankSubtype == 'credit') {
        f.creditCardLimit = v.creditCardLimit.trim().isEmpty
            ? null
            : v.creditCardLimit.trim();
        f.dueDayOfMonth = v.dueDayOfMonth.isEmpty
            ? null
            : int.tryParse(v.dueDayOfMonth);
      }
    case AccountType.brokerage:
      f.openingBalance = v.openingBalance.trim();
      f.investmentType = v.investmentType.isEmpty ? null : v.investmentType;
    case AccountType.gold:
      f.openingBalance = '0';
      f.metalType = v.metalType.isEmpty ? null : v.metalType;
    case AccountType.realEstate:
      f.openingBalance = '0';
      f.propertyType = v.propertyType.isEmpty ? null : v.propertyType;
      f.ownershipPercentage = v.ownershipPercentage.trim();
      f.location = v.location.trim().isEmpty ? null : v.location.trim();
    case AccountType.business:
      f.openingBalance = '0';
      f.businessType = storeClassification(v.businessType, v.businessTypeOther);
      f.industry = storeClassification(v.industry, v.industryOther);
      f.ownershipPercentage = v.ownershipPercentage.trim();
  }
  return f;
}
